plugins/jabra-gnp/fu-jabra-gnp-device.c

```diff
@@ -0,0 +1,859 @@
+	FuJabraGnpVersionData requested_version_data;
+	gchar *requested_version;
+
+	fw_xml =
+	    fu_archive_firmware_get_image_fnmatch(FU_ARCHIVE_FIRMWARE(firmware), "info.xml", error);
+	if (fw_xml == NULL)
+		return FALSE;
+
+	data = fu_firmware_get_bytes(fw_xml, error);
+	if (data == NULL)
+		return FALSE;
+
+	if (!xb_builder_source_load_bytes(source, data, XB_BUILDER_SOURCE_FLAG_NONE, error))
+		return FALSE;
+
+	xb_builder_import_source(builder, source);
+	silo = xb_builder_compile(builder, XB_BUILDER_COMPILE_FLAG_NONE, NULL, error);
+	if (silo == NULL)
+		return FALSE;
+
+	if (!fu_jabra_gnp_device_get_data_from_info(self,
+						    silo,
+						    requested_version,
+						    &requested_version_data,
+						    &info_dfu_pid,
+						    error))
+		return FALSE;
+	files = xb_silo_query(silo, "buildVector/files/file", 0, error);
+	if (files == NULL)
+		return FALSE;
+
+	for (guint i = 0; i < files->len; i++) {
+		XbNode *node = g_ptr_array_index(files, i);
+		FuJabraGnpPartitionData part_data;
+		if (!fu_jabra_gnp_device_get_part_data_from_file_node(node, &part_data, error))
+			return FALSE;
+		if (g_strcmp0(part_data.language, "English") == 0) {
+			fw_tmp =
+			    fu_archive_firmware_get_image_fnmatch(FU_ARCHIVE_FIRMWARE(firmware),
+								  part_data.file_name,
+								  error);
+			data_tmp = fu_firmware_get_bytes(fw_tmp, error);
+			gsize data_len;
+
+			const guint8 *data = g_bytes_get_data(data_tmp, &data_len);
+			guint64 calc_crc = fu_jabra_gnp_device_calculate_crc(data, &data_len);
+			if (calc_crc != part_data.crc)
+				return FALSE;
+			guint chunk_size = 52;
+			g_autoptr(GPtrArray) chunks = NULL;
+			g_autoptr(GBytes) blob = NULL;
+
+			blob = fu_firmware_get_bytes(fw_tmp, error);
+			if (blob == NULL)
+				return FALSE;
+			if (!fu_jabra_gnp_device_write_partition(self, part_data.partition, error))
+				return FALSE;
+			if (!fu_jabra_gnp_device_start(self, error))
+				return FALSE;
+			if (!fu_jabra_gnp_device_flash_erase_done(self, error))
+				return FALSE;
+
+			calc_crc = ((calc_crc & 0xffff) << 16) | ((calc_crc & 0xffff0000) >> 16);
+			chunks = fu_chunk_array_new_from_bytes(blob, 0x00, 0x00, chunk_size);
+			if (!fu_jabra_gnp_device_write_crc(self, calc_crc, chunks->len, 100, error))
+				return FALSE;
+			if (!fu_jabra_gnp_device_write_chunks(self, chunks, error))
+				return FALSE;
+			if (!fu_jabra_gnp_device_read_verify_status(self, error))
+				return FALSE;
+			if (!fu_jabra_gnp_device_write_version(self,
+							       requested_version_data.version_major,
+							       requested_version_data.version_minor,
+							       requested_version_data.version_micro,
+							       error))
+				return FALSE;
+		}
+	}
+	if (!fu_jabra_gnp_device_write_dfu_from_squif(self, error))
+		return FALSE;
+	return TRUE;
+}
+
+static void
+fu_jabra_gnp_device_init(FuJabraGnpDevice *self)
+{
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_protocol(FU_DEVICE(self), "org.jabra.gnp");
```

Comment:
you mean `com.jabra.gnp` -- right?

Reply:
I think you are right, `com.jabra.gnp` makes way more sense. I will need to update our cabinet files, but that shouldn't be an issue

Reply:
I mean the footer-us_en-ev275.gnv file -- is that signed by some Jabra private key or could I flash modified content onto the device?

Reply:
I mean app_p0-us_en-ev275.gnv -- but you get the idea.

Reply:
Our firmware files are signed, I'm just going to get some confirmation on how exactly. If knowing they are signed is enough, the answer should be yes, if you need to know a bit more I'll have to get back to you at a later point

Reply:
"signed" is enough for fwupd -- but the (confidential) detail is needed by the LVFS. You can sent that to me privately if you like.

Reply:
Perfect, I should get confirmation from the team about it and will reach out to you privately.
Just a note that Thursday the 18th and Friday the 19th (for the most part) are public holidays in Denmark, so I won't be working, and will get back to you early next week most likely.